File: src/main/java/com/dyonovan/researchsystem/ResearchSystem.java
package com.dyonovan.researchsystem;

import com.dyonovan.researchsystem.common.CommonProxy;
import com.dyonovan.researchsystem.lib.Reference;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.init.Items;
import net.minecraft.item.ItemStack;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.Mod.EventHandler;
import net.minecraftforge.fml.common.SidedProxy;
import net.minecraftforge.fml.common.event.FMLInitializationEvent;
import net.minecraftforge.fml.common.event.FMLPostInitializationEvent;
import net.minecraftforge.fml.common.event.FMLPreInitializationEvent;
import net.minecraftforge.fml.common.event.FMLServerStartingEvent;

/**
 * This file was created for ResearchSystem
 * <p>
 * ResearchSystem is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Dyonovan
 * @since 12/22/2016
 */
@Mod(modid = Reference.MOD_ID, name = Reference.MOD_NAME, version = Reference.VERSION, dependencies = Reference.DEPENDENCIES, updateJSON = Reference.UPDATE_JSON)
public class ResearchSystem {

    @SidedProxy(clientSide = "com.dyonovan.researchsystem.client.ClientProxy", serverSide = "com.dyonovan.researchsystem.common.CommonProxy")
    public static CommonProxy proxy;

    public static CreativeTabs tabResearchSystem = new CreativeTabs(Reference.MOD_ID + ".tabResearchSystem") {
        @Override
        @SuppressWarnings("NullableProblems")
        public ItemStack getTabIconItem() {
            return new ItemStack(Items.NETHERBRICK);
        }
    };

    @EventHandler
    public void preInit(FMLPreInitializationEvent event) {

    }

    @EventHandler
    public void init(FMLInitializationEvent event) {

    }

    @EventHandler
    public void postInit(FMLPostInitializationEvent event) {

    }

    @EventHandler
    public void serverLoad(FMLServerStartingEvent event) {

    }
}

File: src/main/scala/com/dyonovan/researchsystem/ResearchSystem.scala
package com.dyonovan.researchsystem

import java.io.File

import com.dyonovan.researchsystem.capability.ResearchCapability
import com.dyonovan.researchsystem.common.CommonProxy
import com.dyonovan.researchsystem.common.commands.UnlockCommand
import com.dyonovan.researchsystem.events.CraftingCatcher
import com.dyonovan.researchsystem.lib.Reference
import com.dyonovan.researchsystem.managers.{BlockManager, ConfigManager, EventManager, ResearchManager}
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.init.Blocks
import net.minecraft.item.Item
import net.minecraftforge.fml.common.Mod.EventHandler
import net.minecraftforge.fml.common.event.{FMLInitializationEvent, FMLPostInitializationEvent, FMLPreInitializationEvent, FMLServerStartingEvent}
import net.minecraftforge.fml.common.registry.GameRegistry
import net.minecraftforge.fml.common.{Mod, SidedProxy}
import net.minecraftforge.oredict.RecipeSorter

/**
  * This file was created for ResearchSystem
  *
  * ResearchSystem is licensed under the
  * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
  * http://creativecommons.org/licenses/by-nc-sa/4.0/
  *
  * @author Dyonovan
  * @since 7/22/2016
  */
@Mod(modid = Reference.MOD_ID, name = Reference.MOD_NAME, version = Reference.VERSION, dependencies = Reference.DEPENDENCIES, modLanguage = "scala", updateJSON = Reference.UPDATE_JSON)
object ResearchSystem {

    var configDir: String = ""

    @SidedProxy(clientSide = "com.dyonovan.researchsystem.client.ClientProxy", serverSide = "com.dyonovan.researchsystem.common.CommonProxy")
    var proxy: CommonProxy = _

    var tabResearchSystem: CreativeTabs = new CreativeTabs(Reference.MOD_ID + ".tabResearchSystem") {
        @SuppressWarnings(Array("NullableProblems"))
        def getTabIconItem: Item = Item.getItemFromBlock(Blocks.NETHERRACK) //TODO switch icon
    }

    @EventHandler
    def preInit(event: FMLPreInitializationEvent) {
        configDir = event.getModConfigurationDirectory.getAbsolutePath + File.separator + "ResearchSystem"
        ConfigManager.preInit()
        BlockManager.preInit()
        proxy.preInit()
        ResearchCapability.register()
    }

    @EventHandler
    def init(event: FMLInitializationEvent) {
        ResearchManager.init()
        EventManager.init()
    }

    @EventHandler
    def postInit(event: FMLPostInitializationEvent) {
        ResearchManager.removeRecipes()
        GameRegistry.addRecipe(new CraftingCatcher)
        RecipeSorter.register("Testing", classOf[CraftingCatcher], RecipeSorter.Category.SHAPED, "before:minecraft:shaped")
    }

    @EventHandler def serverLoad(event: FMLServerStartingEvent) {
        event.registerServerCommand(new UnlockCommand)
    }
}
